src/lib/lib.jl

```diff
@@ -210,6 +210,42 @@ end
 @generated pair(::Val{k}, v, _=nothing) where k = :($k = v,)
 @generated pair(::Val{k}, v, ::NamedTuple{keys}) where {k,keys} = k isa Int ? :($(getfield(keys, k)) = v,) : :($k = v,)
 
+# ugly hack to make differentiating `getproperty` infer a lot better
+@generated function _pullback(cx::AContext, ::typeof(literal_getproperty), x, ::Val{f}) where f
+    sig(x) = Tuple{x, typeof(f)}
+    pb_sig(x) = Tuple{cx, typeof(getproperty), x, typeof(f)}
+
+    # either `getproperty` has a custom implementation or `_pullback(cx, ::getproperty, x, f)`
+    # is overloaded directly
+    is_getfield_fallback = which(_pullback, pb_sig(x)) == which(_pullback, pb_sig(Any)) && 
+        which(getproperty, sig(x)) == which(getproperty, sig(Any))
+
+    if is_getfield_fallback
+        # just copy pullback of `literal_getfield`
+        _sig = Tuple{cx, typeof(literal_getfield), x, Val{f}}
+        ci = code_lowered(_pullback, _sig)[1]
+
+        Meta.partially_inline!(
+            ci.code, Any[_pullback, Core.SlotNumber(2), literal_getfield],
+            Tuple{typeof(_pullback), _sig.parameters...}, Any[_sig.parameters...],
+            0, 0, :propagate,
+        )
+
+        # backedge for `getproperty`
```

Comment:
Possibly a silly suggesion, but maybe it would be a good idea to link to the docs regarding backedges? e.g. [here](https://docs.julialang.org/en/v1/devdocs/ast/#MethodInstance) They're definitely one of the less well-known features, and this might this kind of code more accessible to the uninitiated.